lib/libutee/tee_api_operations.c

```diff
@@ -1224,46 +1273,69 @@ TEE_Result TEE_AEDecryptFinal(TEE_OperationHandle op,
 	uint32_t tmp_dlen;
 	size_t req_dlen;
 
-	if (op == TEE_HANDLE_NULL || (srcData == NULL && srcLen != 0) ||
-	    destLen == NULL || (destData == NULL && *destLen != 0) ||
-	    (tag == NULL && tagLen != 0))
-		TEE_Panic(0);
-	if (op->info.operationClass != TEE_OPERATION_AE)
-		TEE_Panic(0);
-	if ((op->info.handleState & TEE_HANDLE_FLAG_INITIALIZED) == 0)
-		TEE_Panic(0);
+	if (operation == TEE_HANDLE_NULL ||
+	    (srcData == NULL && srcLen != 0) ||
+	    destLen == NULL ||
+	    (destData == NULL && *destLen != 0) ||
+	    (tag == NULL && tagLen != 0)) {
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
+	}
+
+	if (operation->info.operationClass != TEE_OPERATION_AE) {
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
+	}
+
+	if ((operation->info.handleState & TEE_HANDLE_FLAG_INITIALIZED) == 0) {
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
+	}
 
 	/*
 	 * Check that required destLen is big enough before starting to feed
 	 * data to the algorithm. Errors during feeding of data are fatal as we
 	 * can't restore sync with this API.
 	 */
-	req_dlen = op->buffer_offs + srcLen;
+	req_dlen = operation->buffer_offs + srcLen;
 	if (*destLen < req_dlen) {
 		*destLen = req_dlen;
-		return TEE_ERROR_SHORT_BUFFER;
+		res = TEE_ERROR_SHORT_BUFFER;
+		goto out;
 	}
 
 	tmp_dlen = *destLen - acc_dlen;
-	tee_buffer_update(op, utee_authenc_update_payload, srcData, srcLen,
-			  dst, &tmp_dlen);
+	res = tee_buffer_update(operation, utee_authenc_update_payload, srcData,
+				srcLen, dst, &tmp_dlen);
+	if (res != TEE_SUCCESS)
+		goto out;
+
 	dst += tmp_dlen;
 	acc_dlen += tmp_dlen;
 
 	tmp_dlen = *destLen - acc_dlen;
-	res =
-	    utee_authenc_dec_final(op->state, op->buffer, op->buffer_offs, dst,
-				   &tmp_dlen, tag, tagLen);
-	if (res != TEE_SUCCESS && res != TEE_ERROR_MAC_INVALID)
-		TEE_Panic(res);
+	res = utee_authenc_dec_final(operation->state, operation->buffer,
+				     operation->buffer_offs, dst, &tmp_dlen,
+				     tag, tagLen);
+	if (res != TEE_SUCCESS)
+		goto out;
+
 	/* Supplied tagLen should match what we initiated with */
-	if (tagLen != op->ae_tag_len)
+	if (tagLen != operation->ae_tag_len) {
 		res = TEE_ERROR_MAC_INVALID;
+		goto out;
```

Comment:
The operation is complete, there's nothing left to do with the operation handle in that state (how can the application know if the tag was truly invalid or just a tagLen mismatch?). It would be an error of the application to call this function again in that state.

*The operation may be in either initial or active state and enters initial state afterwards.*

What isn't clear about that is what to do when returning `TEE_ERROR_SHORT_BUFFER`, that error code doesn't make sense unless we allow the function to be called again.